Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationWindow.cs

```diff
@@ -0,0 +1,417 @@
+            "Objects",
+            "Scenes",
+            "Full Project"
+        };
+
+        private ToolbarOption selectedToolbar = ToolbarOption.Objects;
+        private Vector2 scrollPosition = Vector2.zero;
+
+        private const string DependencyWindowURL = "https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Tools/MigrationWindow.html";
+        private const string WindowTitle = "Migration Window";
+        private const string WindowDescription = "This tool allows the migration of obsolete components into up-to-date versions. Component-specific implementation of the MigrationHandler Interface is required.";
+
+        private const string SceneExtension = ".unity";
+        private const string PrefabExtension = ".prefab";
+
+        private Dictionary<string, Type> migrationTypesMap = new Dictionary<string, Type>();
+        private String[] migrationTypeNames;
+        private int selectedMigrationHandlerIndex;
+        private List<Object> migrationObjects = new List<Object>();
+        private Object selection;
+
+        private dynamic migrationTypeInstance
+        {
+            get
+            {
+                var selectedMigrationTypeName = migrationTypeNames[selectedMigrationHandlerIndex];
+                var migrationHandlerType = migrationTypesMap[selectedMigrationTypeName];
+                return Activator.CreateInstance(migrationHandlerType) as IMigrationHandler;
+            }
+        }
+
+        [MenuItem("Mixed Reality Toolkit/Utilities/Migration Window", false, 4)]
+        private static void ShowWindow()
+        {
+            var window = GetWindow<MigrationWindow>(typeof(SceneView));
+            window.titleContent = new GUIContent(WindowTitle, EditorGUIUtility.IconContent("d_TimelineEditModeRippleOFF").image);
+            window.minSize = new Vector2(400.0f, 300.0f);
+            window.Show();
+        }
+
+        private void OnGUI()
+        {
+            if (EditorApplication.isPlaying || EditorApplication.isPaused)
+            {
+                GUI.enabled = false;
+            }
+            DrawHeader();
+            DrawMigrationTypeSelector();
+            DrawMigrationToolbars();
+        }
+
+        private void OnFocus()
+        {
+            RefreshAvailableTypes();
+        }
+
+        private void RefreshAvailableTypes()
+        {
+            var type = typeof(IMigrationHandler);
+            var types = AppDomain.CurrentDomain.GetAssemblies()
+                .SelectMany(x => x.GetTypes())
+                .Where(x => type.IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract).ToList();
+
+            foreach (var migrationHandlerType in types)
+            {
+                if (!migrationTypesMap.ContainsKey(migrationHandlerType.Name))
+                {
+                    migrationTypesMap.Add(migrationHandlerType.Name, migrationHandlerType);
+                }
+            }
+            migrationTypeNames = new String[migrationTypesMap.Count + 1];
```

Comment:
Can you leave a comment why this is +1ed here (and below?)

I assume this is because Unity gives us 1-indexed selected values, but this isn't necessarily immediately obvious to mere mortals (i.e. myself)

Reply:
Matching popup type name selector  with the actual type list.